### PR TITLE
chore: upgrade to jose v4 library

### LINKED ISSUE #797: Consider upgrading to github.com/go-jose/go-jose/v4
### Preflight checklist

- [X] I could not find a solution in the existing issues, docs, nor discussions.
- [X] I agree to follow this project's [Code of Conduct](https://github.com/ory/fosite/blob/master/CODE_OF_CONDUCT.md).
- [X] I have read and am following this repository's [Contribution Guidelines](https://github.com/ory/fosite/blob/master/CONTRIBUTING.md).
- [ ] I have joined the [Ory Community Slack](https://slack.ory.sh).
- [ ] I am signed up to the [Ory Security Patch Newsletter](https://ory.us10.list-manage.com/subscribe?u=ffb1a878e4ec6c0ed312a3480&id=f605a41b53).

### Ory Network Project

_No response_

### Describe your problem

`github.com/go-jose/go-jose/v3` dependency has made a new `github.com/go-jose/go-jose/v4` version. It breaks backwards compatibility to [improve security](https://github.com/go-jose/go-jose/blob/main/CHANGELOG.md):

> This release makes some breaking changes in order to more thoroughly address the vulnerabilities discussed in [Three New Attacks Against JSON Web Tokens](https://i.blackhat.com/BH-US-23/Presentations/US-23-Tervoort-Three-New-Attacks-Against-JSON-Web-Tokens.pdf), "Sign/encrypt confusion", "Billion hash attack", and "Polyglot token".

I think it is not critical, but it would be beneficial to do so sooner than later.

### Describe your ideal solution

We upgrade.

### Workarounds or alternatives

We do not.

### Version

latest master

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by mitar**
### Related Issue or Design Document

Fixes https://github.com/ory/fosite/issues/797.

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md) and signed the CLA.
- [x] I have referenced an issue containing the design document if my change introduces a new feature.
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security vulnerability. 
      If this pull request addresses a security vulnerability, 
      I confirm that I got approval (please contact [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push the changes.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added the necessary documentation within the code base (if appropriate).

## Further comments

This is just a formality upgrade to v4 of the library, but it does not really take advantage of the new features v4 support. For that I think we should have a more thorough rewrite of current codebase related to JWTs. I was looking into that but it is really hard to do it unless we have clear understanding that this should be done and what can we break in terms of backwards compatibility. Because there are some simple structures we have in fosite/token/jwt package which could probably just be removed and delegated to go-jose directly. I think current fosite/token/jwt was written to be backwards compatible with prior version of fosite/token/jwt which used a much smaller jwt library so a lot of logic had to be done in fosite/token/jwt. But now fosite/token/jwt is doing very strange loops where it duplicates a lot of structs and work of go-jose. For example, fosite/token/jwt validates the token using go-jose (using all reasonable algorithms) and then checks if it matches the expected signing algorithm. It would be much better if we would just give to go-jose the signing algorithm which we expect in the first place and then we do not have to do no checks anymore in fosite/token/jwt (which was the reason for [v4 in the first place](https://github.com/go-jose/go-jose/issues/64#issuecomment-1809152412)). Similarly with private/public keys. We do a lot of work where we take private/public keys and inspect them and reconstruct them and then pass them to go-jose. But go-jose already supports all of that and you can already pass keys directly (which is what https://github.com/ory/fosite/pull/799 was about), even more, you can pass whole key sets directly and it picks the correct key (which fosite also tries to do). I also think fosite is probably [susceptible to this attack](https://auth0.com/blog/critical-vulnerabilities-in-json-web-token-libraries/) with current approach of just blindly validating tokens and only then (sometimes) checking if signing algorithm is what was expected (I see `GetRequestObjectSigningAlgorithm` and `GetTokenEndpointAuthSigningAlgorithm`, but nothing to check when validating tokens themselves, like in introspect endpoint). Maybe I am missing something.

TLDR; I think this upgrade is the first step, but somebody with executive permissions to break backwards compatibility should give some love and cleanup current jwt codebase in fosite.

### CODE REVIEW

**Comment by mitar** [comment]
(CI is failing because Hydra has not been upgraded to jose v4.)

File: token/jwt/token.go [hunk 1]
```diff
@@ -37,6 +37,12 @@ const (
 	JWTHeaderTypeValue = "JWT"
 )
 
+var SupportedSignatureAlgorithms = []jose.SignatureAlgorithm{
+	SigningMethodNone,
```
Comment (aeneasr) [inline token/jwt/token.go:41]: Could we have missed some here that would cause some breaking changes in existing systems because the validation has changed?

**Review by aeneasr — COMMENTED** [review]
One question, otherwise looks good to me :)

File: token/jwt/token.go (hunk 1 above)
Comment (mitar) [reply token/jwt/token.go:41]: I listed everything standardized and available and secure. In the worst case though, this is a publicly accessible variable in Go, so people can extend it if we did miss something. Not nice API though.

But as I mentioned in my PR description, I really think we should do a backwards incompatible change after this, to address current code rot around JWTs in fosite and address some vulnerabilities available exactly because we allow for example to try all algorithms and not just the one expected. But this is probably something only you can decide on how much breaking change would you approve (not in terms of external users changes, but in terms of API changes for fosite integrators - so some fosite structs we could just get rid of and delegate to go-jose).

Comment (aeneasr) [reply token/jwt/token.go:41]: I'm pretty sure we validate the signing method somewhere, do we not?

**Comment by aeneasr** [comment]
>I also think fosite is probably [susceptible to this attack](https://auth0.com/blog/critical-vulnerabilities-in-json-web-token-libraries/) with current approach of just blindly validating tokens and only then (sometimes) checking if signing algorithm is what was expected (I see GetRequestObjectSigningAlgorithm and GetTokenEndpointAuthSigningAlgorithm, but nothing to check when validating tokens themselves, like in introspect endpoint). Maybe I am missing something.

I did take a look into this. The biggest issue of the attack is using alg:"none" which we support. However, to pass, the key func must return a constant type `UnsafeAllowNoneSignatureType` which we use only in `authorizeRequestParametersFromOpenIDConnectRequest`.

So at least from that perspective we're safe. However, I totally agree with you that the JWT code is horrific and needs some good refactoring and simplification. I agree that backwards compatibility will be quite challenging but we can probably test for it?

**Review by aeneasr — APPROVED** [review]

**Comment by mitar** [comment]
> However, I totally agree with you that the JWT code is horrific and needs some good refactoring and simplification. I agree that backwards compatibility will be quite challenging but we can probably test for it?

What I meant is that it might beneficial to drop backwards compatibility here and just get rid of the whole jwt fosite package and document how any users of the code should migrate to directly use go-jose v4. Probably it is possible to keep backwards compatibility, but it would require us to a) maintain multiple structs and copy data between them b) probably still change some of the methods, like to pass now an expected sign algorithm.

**Comment by aeneasr** [comment]
> What I meant is that it might beneficial to drop backwards compatibility here and just get rid of the whole jwt fosite package and document how any users of the code should migrate to directly use go-jose v4.

API BC breaks are totally fine with me, breaking existing implementations / expectations on the API behavior is not (because we'll be busy with customer support for a week if that happens :D)

**Comment by aeneasr** [comment]
There's a couple of failing things and merge conflicts. Would you mind taking a look? It would be good to get this merged

**Comment by mitar** [comment]
Yes, I can check.

**Comment by sword-jin** [comment]
@mitar 
thanks for your great work.
I'd like to have a friendly ping, I am also very looking forward this PR.

**Comment by mitar** [comment]
I still have this on my radar, but I am currently swamped with some other work. I can give anyone access to my branch, or feel free to fork my branch and just do another PR with updated PR? I do not mind. Otherwise, I hope I will be able to get to this at the end of this month, but we will see.

**Comment by mitar** [comment]
I refreshed the PR with the latest master branch. Fosite tests are passing. Only Hydra tests are not because Hydra itself will have to be upgraded to v4 version of the jose library. Not sure what is the process of getting them to pass here? It is a chicken and an egg problem?

**Comment by james-d-elliott** [comment]
You can make a similar PR in hydra that updates to this commit sha, and then reference that commit sha here for tests.

**Comment by mitar** [comment]
@aeneasr How do I go about tests here?

**Comment by mitar** [comment]
@aeneasr Please advise here how do I go about tests here.

**Comment by aeneasr** [comment]
Sorry, I missed that. You will need to create a PR on hydra, and use a go mod rewrite to use your fork and git commit as a dependency for fosite.

Then here in fosite, you need to change https://github.com/ory/fosite/actions/runs/15810823776/workflow?pr=824#L16-L17 (which I just saw is still pointing to canonical ;) ) to your fork of hydra and git commit / branch

**Comment by mitar** [comment]
@aeneasr Tests now pass. Please merge.

Also see corresponding MR in Hydra: https://github.com/ory/hydra/pull/4032

**Comment by clifg** [comment]
Thanks for your work here, @mitar ! @aeneasr , is this change good to merge and if not do you have any other changes/fixes you'd need in order to merge? I'd like to resolve our code's govulncheck errors due to the indirect usage of `go-jose/v3`. Thanks!

**Comment by clifg** [comment]
@aeneasr - Will the maintainers be able to merge this request? We're starting to have vulnchecks pile up against fosite and some of them, like this one, have fixes ready from contributors.

**Comment by clifg** [comment]
@aeneasr - Is the fosite library still being supported? The last update was over a year ago and contributors have open PRs that aren't reviewed or merged. We'd like to get this go-jose update in, and also contribute some other changes (we'd like to see if CIMD support can be added!) but I'm not sure if we should try to build things here or just start maintaining our own fork 😅
